From a4e971ae8a5632568839adcd905db191585694a6 Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Fri, 1 Jan 2021 03:07:52 +0530
Subject: [PATCH] libcameraservice: Add support for miui camera mode

 * devices like ginkgo and some xiaomi sdm660 use miui camera mode in camera
   hal to activate certain functions in camera hal, these are enabled when
   vendor.camera.miui.apk is set to 1 based on sys.camera.miui.apk value

 * if this prop is set by default gcam crashes, so we must do it dynamically

 * xiaomi does this in stock libcameraservice but unfortunately we don't
   have stock android 12 to use prebuilt lib

Change-Id: I8d9ee4e650f3e2196546570c183c9d169b8aa335
Signed-off-by: Joey Huab <joey@evolution-x.org>
Signed-off-by: bhaskar966 <deybhaskar999@gmail.com>
---
 services/camera/libcameraservice/CameraService.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/services/camera/libcameraservice/CameraService.cpp b/services/camera/libcameraservice/CameraService.cpp
index 081ce6c0aa..ee7ea061ae 100644
--- a/services/camera/libcameraservice/CameraService.cpp
+++ b/services/camera/libcameraservice/CameraService.cpp
@@ -36,6 +36,7 @@
 #include <aidl/AidlCameraService.h>
 #include <android-base/macros.h>
 #include <android-base/parseint.h>
+#include <android-base/properties.h>
 #include <android-base/stringprintf.h>
 #include <android/permission/Pecd /path/to/your/directory
for file in *; do
    mv "$file" "${file// /-}"
donermissionChecker.h>
 #include <binder/ActivityManager.h>
@@ -92,6 +93,7 @@ namespace {
 namespace android {
 
 using base::StringPrintf;
+using base::SetProperty;
 using binder::Status;
 using namespace camera3;
 using frameworks::cameraservice::service::V2_0::implementation::HidlCameraService;
@@ -3901,6 +3903,15 @@ status_t CameraService::BasicClient::startCameraOps() {
 
     mOpsActive = true;
 
+    // Configure miui camera mode
+    if (strcmp(String8(mClientPackageName).string(), "com.android.camera") == 0) {
+        SetProperty("sys.camera.miui.apk", "1");
+        ALOGI("Enabling miui camera mode");
+    } else {
+        SetProperty("sys.camera.miui.apk", "0");
+        ALOGI("Disabling miui camera mode");
+    }
+
     // Transition device availability listeners from PRESENT -> NOT_AVAILABLE
     sCameraService->updateStatus(StatusInternal::NOT_AVAILABLE, mCameraIdStr);
 
